README.md
# GeoRSSy
[![PyPI](https://img.shields.io/pypi/v/georssy.svg)](https://pypi.python.org/pypi/georssy)
[![PyPI](https://img.shields.io/pypi/status/georssy.svg)](https://pypi.python.org/pypi/georssy)
[![PyPI](https://img.shields.io/pypi/l/georssy.svg)](https://pypi.python.org/pypi/georssy)
[![PyPI](https://img.shields.io/pypi/pyversions/georssy.svg)](https://pypi.python.org/pypi/georssy)
[![PyPI](https://img.shields.io/pypi/format/georssy.svg)](https://pypi.python.org/pypi/georssy)
[![PyPI](https://img.shields.io/pypi/wheel/georssy.svg)](https://pypi.python.org/pypi/georssy)
[![PyPI](https://img.shields.io/badge/Say%20Thanks-!-1EAEDB.svg)](https://saythanks.io/to/devsf)

A rough Python GeoRSS (Geographically Encoded Objects for RSS feeds) decoder.

GeoRSSy is a very rough (but powerfull!) decoder for the GeoRSS (Geographically Encoded Objects for RSS feeds) standard.

For information about GeoRSS: http://www.georss.org/

## How to install
To install GeoRSSy, simply:
```bash
$ pip install georssy
```

## How to use
To use georss, simply:
```python
from georssy import GeoRssDecoder
...
georss_entry = GeoRssDecoder( parent_node = r, polygons_over_boxes = True )
tmp = georss_entry.polygon_list
```

## How to Contribute
1. Check for open issues or open a fresh issue to start a discussion around a feature idea or a bug.
2. Fork the repository on GitHub to start making your changes to the master branch (or branch off of it).
3. Write a test which shows that the bug was fixed or that the feature works as expected.
4. Send a pull request and ask the maintainer to merge and publish it.

## TODO
GeoRSSy is a very young project, so there's a lot of things to do! This is a short TO-DO list:
- accept URL to a remote/local XML as input files
- GML: add "Coordinate Reference System" attributes implementation (and add new tests)
- check Python 3 compatibility

---
# WARNING!!! GeoRSSy is currently under development! So it's not so stable as you can expect! :)
---


HISTORY.rst
.. :changelog:

Release History
---------------

v0.0.3 (2017-08-25)
++++++++++++++++

- Bug fixing
- Added Tests
- Created Makefile

v0.0.2 (2017-08-24)
++++++++++++++++

- PyPi improvements

dev (2017-08-24)
++++++++++++++++

- PyPi improvements

v0.0.1 (2017-08-24)
++++++++++++++++

- First Release

dev (2017-08-24)
++++++++++++++++

- Added GeoRSSEntry model class
- Created "Tests" folder and moved Test.py into it
- Added "decoder" functions into simple_decoder.py and gml_decoder.py. Changed decoder.py implementation
- Added api.py module
- Prepared for pip release

dev (2017-08-24)
++++++++++++++++

- First Commit


georssy/decoder.py
"""
georssy.decoder
~~~~~~~~~~~~
TODO
:copyright: (c) 2017 by Sergio Ferraresi.
:license: Apache2, see LICENSE for more details.
"""

import logging
import re

from .gml.gml_decoder import gml_decoder
from .simple.simple_decoder import simple_decoder

logger = logging.getLogger( __name__ )

class decoder( object ):
    '''
    GeoRss Parser.
    GeoRSS-Simple is meant as a very lightweight format that developers and users can quickly and easily add to their existing feeds with little effort. It supports basic geometries (point, line, box, polygon) and covers the typical use cases when encoding locations.
    GeoRSS GML is a formal GML Application Profile, and supports a greater range of features, notably coordinate reference systems other than WGS-84 latitude/longitude.

    Some publishers and users may prefer to separate lat/long pairs by a comma rather than whitespace. This is permissible in Simple; GeoRSS parsers should just treat commas as whitespace.

    See http://www.georss.org/ for more information.
    '''

    def __init__( self, parent_node = None, polygons_over_boxes = False ):
        '''
        Constructor
        '''

        if parent_node is None:
            msg = 'GeoRSS parent node NOT valid'
            logger.error( msg )

            raise ValueError( msg )

        logger.debug( 'georssy parameters:' )
        logger.debug( '  Polygons over Boxes: "%s"' % ( 'Y' if polygons_over_boxes else 'N' ) )

        sd = simple_decoder( parent_node = parent_node, polygons_over_boxes = polygons_over_boxes )
        gd = gml_decoder( parent_node = parent_node, polygons_over_boxes = polygons_over_boxes )

        self.point_list = []
        if sd.point_list:
            self.point_list.extend( sd.point_list )
        if gd.point_list:
            self.point_list.extend( gd.point_list )
        self.line_list = []
        if sd.line_list:
            self.line_list.extend( sd.line_list )
        if gd.line_list:
            self.line_list.extend( gd.line_list )
        self.polygon_list = []
        if sd.polygon_list:
            self.polygon_list.extend( sd.polygon_list )
        if gd.polygon_list:
            self.polygon_list.extend( gd.polygon_list )
        self.feature_type_list = sd.feature_type_list
        self.feature_name_list = sd.feature_name_list
        self.relationship_list = sd.relationship_list
        self.elevation_list = sd.elevation_list
        self.floor_list = sd.floor_list
        self.radius_list = sd.radius_list

        logger.debug( 'georssy decoded elements:' )
        if self.point_list:
            logger.debug( '  Points: "%s"' % str( self.point_list ) )
        if self.line_list:
            logger.debug( '  Lines: "%s"' % str( self.line_list ) )
        if self.polygon_list:
            logger.debug( '  Polygons: "%s"' % str( self.polygon_list ) )
        if self.feature_type_list:
            logger.debug( '  Feature Types: "%s"' % str( self.feature_type_list ) )
        if self.feature_name_list:
            logger.debug( '  Feature Names: "%s"' % str( self.feature_name_list ) )
        if self.relationship_list:
            logger.debug( '  Relationships: "%s"' % str( self.relationship_list ) )
        if self.elevation_list:
            logger.debug( '  Elevations: "%s"' % str( self.elevation_list ) )
        if self.floor_list:
            logger.debug( '  Floors: "%s"' % str( self.floor_list ) )


georssy/simple/simple_decoder.py
"""
georssy.simple.simple_decoder
~~~~~~~~~~~~
TODO
:copyright: (c) 2017 by Sergio Ferraresi.
:license: Apache2, see LICENSE for more details.
"""

import logging
import re

logger = logging.getLogger( __name__ )

#def decoder( parent_node ):
#    sd = simple_decoder( parent_node = parent_node )
#
#    return [ sd.point_list, sd.line_list, sd.polygon_list, sd.feature_type_list, sd.feature_name_list, sd.relationship_list, sd.elevation_list, sd.floor_list, sd.radius_list ]

class simple_decoder( object ):
    '''
    GeoRss Parser.
    GeoRSS-Simple is meant as a very lightweight format that developers and users can quickly and easily add to their existing feeds with little effort. It supports basic geometries (point, line, box, polygon) and covers the typical use cases when encoding locations.
    GeoRSS GML is a formal GML Application Profile, and supports a greater range of features, notably coordinate reference systems other than WGS-84 latitude/longitude.

    Some publishers and users may prefer to separate lat/long pairs by a comma rather than whitespace. This is permissible in Simple; GeoRSS parsers should just treat commas as whitespace.

    See http://www.georss.org/ for more information.
    '''

    def __init__( self, parent_node, polygons_over_boxes = False ):
        '''
        Constructor
        '''

        self.parent_node = parent_node
        self.p_over_b = polygons_over_boxes

        self.point_list = []
        self.line_list = []
        self.polygon_list = []
        self.feature_type_list = []
        self.feature_name_list = []
        self.relationship_list = []
        self.elevation_list = []
        self.floor_list = []
        self.radius_list = []

        self.__georss_simple()

        # Common
        self.__feature_types()
        self.__feature_name()
        self.__relationships()
        self.__elevations()
        self.__floors()
        self.__radiuses()

    def __georss_simple( self ):
        self.__simple_points()
        self.__simple_lines()
        if self.p_over_b:
            self.__simple_polygon()
        else:
            self.__simple_polygon()
            self.__simple_box()
        # TODO CIRCLE

    def __simple_points( self ):
        '''
        Example: <georss:point>45.256 -71.92</georss:point>
        '''

        point_list = self.parent_node.findall( '{http://www.georss.org/georss}point' )
        if point_list:
            logger.debug( '  "Simple Point" detected! Decoding...' )
            for point in point_list:
                tmp = re.sub( '\s+', ' ', point.text.strip().replace( ',', ' ' ) ).split()
                self.point_list.append( {
                    'type': 'POINT',
                    'mode': 'latlon',
                    'coordinates': [ float( tmp[ 0 ] ), float( tmp[ 1 ] ) ],
                    'raw': 'POINT(%s %s)' % ( str( float( tmp[ 1 ] ) ), str( float( tmp[ 0 ] ) ) ),
                } )
            return True

        logger.debug( '  No "Simple Point" detected! Nothing to decode.' )
        return False

    def __simple_lines( self ):
        '''
        Example: <georss:line>45.256 -110.45 46.46 -109.48 43.84 -109.86</georss:line>
        '''

        line_list = self.parent_node.findall( '{http://www.georss.org/georss}line' )
        if line_list:
            logger.debug( '  "Simple Line" detected! Decoding...' )
            for line in line_list:
                tmp = re.sub( '\s+', ' ', line.text.strip().replace( ',', ' ' ) ).split()
                # Gel all Latitudes
                lat_list = tmp[ 0 : len( tmp ) : 2 ]
                # Gel all Longitudes
                lon_list = tmp[ 1 : len( tmp ) : 2 ]

                line_tmp = []
                line_raw = ''
                for lat, lon in zip( lat_list, lon_list ):
                    line_tmp.append( [ float( lat ), float( lon ) ] )
                    line_raw += '%s %s,' % ( str( float( lon ) ), str( float( lat ) ) )

                self.line_list.append( {
                    'type': 'LINESTRING',
                    'mode': 'latlon',
                    'coordinates': line_tmp,
                    'raw': 'LINESTRING(%s)' % line_raw[ 0 : -1 ], # remove last comma
                } )
            return True

        logger.debug( '  No "Simple Line" detected! Nothing to decode.' )
        return False

    def __simple_polygon( self ):
        '''
        Example: <georss:polygon>45.256 -110.45 46.46 -109.48 43.84 -109.86 45.256 -110.45</georss:polygon>
        '''

        polygon_list = self.parent_node.findall( '{http://www.georss.org/georss}polygon' )
        if polygon_list:
            logger.debug( '  "Simple Polygon" detected! Decoding...' )
            for polygon in polygon_list:
                tmp = re.sub( '\s+', ' ', polygon.text.strip().replace( ',', ' ' ) ).split()
                # Gel all Latitudes
                lat_list = tmp[ 0 : len( tmp ) : 2 ]
                # Gel all Longitudes
                lon_list = tmp[ 1 : len( tmp ) : 2 ]

                polygon_tmp = []
                polygon_raw = '('
                for lat, lon in zip( lat_list, lon_list ):
                    polygon_tmp.append( [ float( lat ), float( lon ) ] )
                    polygon_raw += '%s %s,' % ( str( float( lon ) ), str( float( lat ) ) )
                # remove last comma
                polygon_raw = polygon_raw[ 0 : -1 ] + ')'

                self.polygon_list.append( {
                    'type': 'POLYGON',
                    'mode': 'latlon',
                    'coordinates': [ polygon_tmp ],
                    'raw': 'POLYGON(%s)' % polygon_raw,
                } )
            return True

        logger.debug( '  No "Simple Polygon" detected! Nothing to decode.' )
        return False

    def __simple_box( self ):
        '''
        Example: <georss:box>42.943 -71.032 43.039 -69.856</georss:box>
        '''

        box_list = self.parent_node.findall( '{http://www.georss.org/georss}box' )
        if box_list:
            logger.debug( '  "Simple Box" detected! Decoding...' )
            for box in box_list:
                tmp = re.sub( '\s+', ' ', box.text.strip().replace( ',', ' ' ) ).split()
                lc = [ tmp[ 0 ], tmp[ 1 ] ]
                uc = [ tmp[ 2 ], tmp[ 3 ] ]

                polygon_tmp = []
                polygon_tmp.append( [ float( uc[ 0 ] ), float( uc[ 1 ] ) ] )
                polygon_tmp.append( [ float( lc[ 0 ] ), float( uc[ 1 ] ) ] )
                polygon_tmp.append( [ float( lc[ 0 ] ), float( lc[ 1 ] ) ] )
                polygon_tmp.append( [ float( uc[ 0 ] ), float( lc[ 1 ] ) ] )
                polygon_tmp.append( [ float( uc[ 0 ] ), float( uc[ 1 ] ) ] )
                raw_tmp = ''
                raw_tmp += '%s %s,' % ( str( float( uc[ 1 ] ) ), str( float( uc[ 0 ] ) ) )
                raw_tmp += '%s %s,' % ( str( float( uc[ 1 ] ) ), str( float( lc[ 0 ] ) ) )
                raw_tmp += '%s %s,' % ( str( float( lc[ 1 ] ) ), str( float( lc[ 0 ] ) ) )
                raw_tmp += '%s %s,' % ( str( float( lc[ 1 ] ) ), str( float( uc[ 0 ] ) ) )
                raw_tmp += '%s %s'  % ( str( float( uc[ 1 ] ) ), str( float( uc[ 0 ] ) ) )
                self.polygon_list.append( {
                    'type': 'POLYGON',
                    'mode': 'latlon',
                    'coordinates': [ polygon_tmp ],
                    'raw': 'POLYGON((%s))' % raw_tmp,
                } )
            return True

        logger.debug( '  No "Simple Box" detected! Nothing to decode.' )
        return False

    def __feature_types( self ):
        '''
        Example: <georss:featuretypetag>city</georss:featuretypetag>
        '''

        feature_type_list = self.parent_node.findall( '{http://www.georss.org/georss}featuretypetag' )
        if feature_type_list:
            logger.debug( '  "Feature Type" detected! Decoding...' )
            for feature_type in feature_type_list:
                tmp = feature_type.text.strip()

                self.feature_type_list.append( {
                    'type': 'FEAT_TYPE',
                    'value': tmp,
                } )
            return True

        logger.debug( '  No "Feature Type" detected! Nothing to decode.' )
        return False

    def __feature_name( self ):
        '''
        Example: <georss:featurename>Podunk</georss:featurename>
        '''

        feature_name_list = self.parent_node.findall( '{http://www.georss.org/georss}featurename' )
        if feature_name_list:
            logger.debug( '  "Feature Name" detected! Decoding...' )
            for feature_name in feature_name_list:
                tmp = feature_name.text.strip()

                self.feature_name_list.append( {
                    'type': 'FEAT_NAME',
                    'value': tmp,
                } )
            return True

        logger.debug( '  No "Feature Name" detected! Nothing to decode.' )
        return False

    def __relationships( self ):
        '''
        Example: <georss:relationshiptag>is-centered-at</georss:relationshiptag>
        '''

        relationship_list = self.parent_node.findall( '{http://www.georss.org/georss}relationshiptag' )
        if relationship_list:
            logger.debug( '  "Relationship" detected! Decoding...' )
            for relationship in relationship_list:
                tmp = relationship.text.strip()

                self.relationship_list.append( {
                    'type': 'RELAT',
                    'value': tmp,
                } )
            return True

        logger.debug( '  No "Relationship" detected! Nothing to decode.' )
        return False

    def __elevations( self ):
        '''
        Example: <georss:elev>313</georss:elev>
        '''

        elevation_list = self.parent_node.findall( '{http://www.georss.org/georss}elev' )
        if elevation_list:
            logger.debug( '  "Elevation" detected! Decoding...' )
            for elevation in elevation_list:
                tmp = elevation.text.strip()

                self.elevation_list.append( {
                    'type': 'ELEV',
                    'value': tmp,
                } )
            return True

        logger.debug( '  No "Elevation" detected! Nothing to decode.' )
        return False

    def __floors( self ):
        '''
        Example: <georss:floor>2</georss:floor>
        '''

        floor_list = self.parent_node.findall( '{http://www.georss.org/georss}floor' )
        if floor_list:
            logger.debug( '  "Floor" detected! Decoding...' )
            for floor in floor_list:
                tmp = floor.text.strip()

                self.floor_list.append( {
                    'type': 'FLOOR',
                    'value': tmp,
                } )
            return True

        logger.debug( '  No "Floor" detected! Nothing to decode.' )
        return False

    def __radiuses( self ):
        '''
        Example: <georss:radius>500</georss:radius>
        '''

        radius_list = self.parent_node.findall( '{http://www.georss.org/georss}radius' )
        if radius_list:
            logger.debug( '  "Radius" detected! Decoding...' )
            for radius in radius_list:
                tmp = radius.text.strip()

                self.radius_list.append( {
                    'type': 'RADIUS',
                    'value': tmp,
                } )
            return True

        logger.debug( '  No "Radius" detected! Nothing to decode.' )
        return False


georssy/gml/gml_decoder.py
"""
georssy.gml.gml_decoder
~~~~~~~~~~~~
TODO
:copyright: (c) 2017 by Sergio Ferraresi.
:license: Apache2, see LICENSE for more details.
"""

import logging
import re

logger = logging.getLogger( __name__ )

class gml_decoder( object ):
    '''
    GeoRss Parser.
    GeoRSS-Simple is meant as a very lightweight format that developers and users can quickly and easily add to their existing feeds with little effort. It supports basic geometries (point, line, box, polygon) and covers the typical use cases when encoding locations.
    GeoRSS GML is a formal GML Application Profile, and supports a greater range of features, notably coordinate reference systems other than WGS-84 latitude/longitude.

    Some publishers and users may prefer to separate lat/long pairs by a comma rather than whitespace. This is permissible in Simple; GeoRSS parsers should just treat commas as whitespace.

    See http://www.georss.org/ for more information.
    '''

    def __init__( self, parent_node, polygons_over_boxes = False ):
        '''
        Constructor
        '''

        self.parent_node = parent_node
        self.p_over_b = polygons_over_boxes
        self.where_node = None

        self.point_list = []
        self.line_list = []
        self.polygon_list = []

        self.__georss_gml()

    def __georss_gml( self ):
        '''
        Example:
        <georss:where>
            <gml:Point>
                <gml:pos>45.256 -71.92</gml:pos>
            </gml:Point>
        </georss:where>
        '''

        self.where_node = self.parent_node.find( '{http://www.georss.org/georss}where' )

        if self.where_node is not None:
            self.__gml_points()
            self.__gml_lines()
            if self.p_over_b:
                self.__gml_polygons()
            else:
                self.__gml_polygons()
                self.__gml_boxes()

    def __gml_points( self ):
        '''
        Example:
            <gml:Point>
                <gml:pos>45.256 -71.92</gml:pos>
            </gml:Point>

        The following are DEPRECATED:
            <gml:Point>
                <gml:coordinates>...</gml:coordinates>
            </gml:Point>

            <gml:Point>
                <gml:coord>...</gml:coord>
            </gml:Point>
        '''

        point_list = self.where_node.findall( '{http://www.opengis.net/gml}Point' )
        if point_list:
            logger.debug( '  "GML Point" detected! Decoding...' )
            for point in point_list:
                pos_list = point.findall( '{http://www.opengis.net/gml}pos' )

                for pos in pos_list:
                    tmp = re.sub( '\s+', ' ', pos.text.strip().replace( ',', ' ' ) ).split()
                    self.point_list.append( {
                        'type': 'POINT',
                        'mode': 'latlon',
                        'coordinates': [ float( tmp[ 0 ] ), float( tmp[ 1 ] ) ],
                        'raw': 'POINT(%s %s)' % ( str( float( tmp[ 1 ] ) ), str( float( tmp[ 0 ] ) ) ),
                    } )
                return True

            logger.debug( '  No "GML Point" detected! Nothing to decode.' )
            return False

    def __gml_lines( self ):
        '''
        Example:
            <gml:LineString>
                <gml:posList>45.256 -110.45 46.46 -109.48 43.84 -109.86</gml:posList>
            </gml:LineString>
        or:
            <gml:LineString>
                <gml:pos>45.256 -110.45</gml:pos>
                <gml:pos>46.46 -109.48</gml:pos>
                <gml:pos>43.84 -109.86</gml:pos>
            </gml:LineString>
        or:
            <gml:LineString>
                <gml:pointProperty>
                    <gml:Point>
                        <gml:pos>45.256 -110.45</gml:pos>
                    </gml:Point>
                    <gml:Point>
                        <gml:pos>46.46 -109.48</gml:pos>
                    </gml:Point>
                    <gml:Point>
                        <gml:pos>43.84 -109.86</gml:pos>
                    </gml:Point>
                </gml:pointProperty>
            </gml:LineString>

        The following are DEPRECATED:
            <gml:LineString>
                <gml:coordinates>...</gml:coordinates>
            </gml:LineString>

            <gml:LineString>
                <gml:coord>...</gml:coord>
                ...
            </gml:LineString>

            <gml:LineString>
                <gml:pointRep>...</gml:pointRep>
            </gml:LineString>
        '''

        line_list = self.where_node.findall( '{http://www.opengis.net/gml}LineString' )
        if line_list:
            logger.debug( '  "GML Line" detected! Decoding...' )
            for line in line_list:
                poslist = line.find( '{http://www.opengis.net/gml}posList' )
                # get all pos nodes, if existing
                pos_list = line.findall( './/{http://www.opengis.net/gml}pos' )

                if poslist:
                    tmp = re.sub( '\s+', ' ', poslist.text.strip().replace( ',', ' ' ) ).split()
                    # Gel all Latitudes
                    lat_list = tmp[ 0 : len( tmp ) : 2 ]
                    # Gel all Longitudes
                    lon_list = tmp[ 1 : len( tmp ) : 2 ]

                    line_tmp = []
                    line_raw = ''
                    for lat, lon in zip( lat_list, lon_list ):
                        line_tmp.append( [ float( lat ), float( lon ) ] )
                        line_raw += '%s %s,' % ( str( float( lon ) ), str( float( lat ) ) )
                    # remove last comma
                    line_raw = line_raw[ 0 : -1 ]

                    self.line_list.append( {
                        'type': 'LINESTRING',
                        'coordinates': line_tmp,
                        'raw': 'LINESTRING(%s)' % line_raw,
                    } )
                elif pos_list and len( pos_list ) > 0:
                    line_tmp = []
                    line_raw = ''
                    for pos in pos_list:
                        tmp = re.sub( '\s+', ' ', pos.text.strip().replace( ',', ' ' ) ).split()
                        line_tmp.append( [ float( tmp[ 0 ] ), float( tmp[ 1 ] ) ] )
                        line_raw += '%s %s,' % ( str( float( tmp[ 1 ] ) ), str( float( tmp[ 0 ] ) ) )
                    # remove last comma
                    line_raw = line_raw[ 0 : -1 ]

                    self.line_list.append( {
                        'type': 'LINESTRING',
                        'mode': 'latlon',
                        'coordinates': line_tmp,
                        'raw': 'LINESTRING(%s)' % line_raw,
                    } )
                return True

            logger.debug( '  No "GML Line" detected! Nothing to decode.' )
            return False

    def __gml_polygons( self ):
        '''
        Example:
            <gml:Polygon>
                <gml:exterior>
                    <gml:LinearRing>
                        <gml:posList>45.256 -110.45 46.46 -109.48 43.84 -109.86 45.256 -110.45</gml:posList>
                    </gml:LinearRing>
                </gml:exterior>
            </gml:Polygon>
        or:
            <gml:Polygon>
                <gml:exterior>
                    <gml:LinearRing>
                        <gml:pos>45.256 -110.45</gml:pos>
                        <gml:pos>46.46 -109.48</gml:pos>
                        <gml:pos>43.84 -109.86</gml:pos>
                        <gml:pos>45.256 -110.45</gml:pos>
                    </gml:LinearRing>
                </gml:exterior>
            </gml:Polygon>
        or:
            <gml:Polygon>
                <gml:exterior>
                    <gml:LinearRing>
                        <gml:pointProperty>
                            <gml:Point>
                                <gml:pos>45.256 -110.45</gml:pos>
                            </gml:Point>
                            <gml:Point>
                                <gml:pos>46.46 -109.48</gml:pos>
                            </gml:Point>
                            <gml:Point>
                                <gml:pos>43.84 -109.86</gml:pos>
                            </gml:Point>
                            <gml:Point>
                                <gml:pos>45.256 -110.45</gml:pos>
                            </gml:Point>
                        </gml:pointProperty>
                    </gml:LinearRing>
                </gml:exterior>
            </gml:Polygon>

        The following are DEPRECATED:
            <gml:Polygon>
                <gml:exterior>
                    <gml:LinearRing>
                        <gml:coordinates>...</gml:coordinates>
                    </gml:LinearRing>
                </gml:exterior>
            </gml:Polygon>

            <gml:Polygon>
                <gml:exterior>
                    <gml:LinearRing>
                        <gml:coord>...</gml:coord>
                    </gml:LinearRing>
                </gml:exterior>
            </gml:Polygon>

            <gml:Polygon>
                <gml:exterior>
                    <gml:LinearRing>
                        <gml:pointRep>...</gml:pointRep>
                        ...
                    </gml:LinearRing>
                </gml:exterior>
            </gml:Polygon>
        '''

        polygon_list = self.where_node.findall( '{http://www.opengis.net/gml}Polygon' )
        if polygon_list:
            logger.debug( '  "GML Polygon" detected! Decoding...' )
            for polygon in polygon_list:
                exterior = polygon.find( '{http://www.opengis.net/gml}exterior' )
                interior = polygon.find( '{http://www.opengis.net/gml}interior' )

                polygon_tmp = []
                raw_tmp = ''
                if exterior:
                    ring = ext_.find( '{http://www.opengis.net/gml}LinearRing' )

                    if ring:
                        poslist = ring.find( '{http://www.opengis.net/gml}posList' )
                        # get all pos nodes, if existing
                        pos_list = ring.findall( './/{http://www.opengis.net/gml}pos' )

                        ext_pol = []
                        ext_raw = '('
                        if poslist:
                            tmp = re.sub( '\s+', ' ', poslist.text.strip().replace( ',', ' ' ) ).split()
                            # Gel all Latitudes
                            lat_list = tmp[ 0 : len( tmp ) : 2 ]
                            # Gel all Longitudes
                            lon_list = tmp[ 1 : len( tmp ) : 2 ]

                            for lat, lon in zip( lat_list, lon_list ):
                                ext_pol.append( [ float( lat ), float( lon ) ] )
                                ext_raw += '%s %s,' % ( str( float( lon ) ), str( float( lat ) ) )
                            # remove last comma
                            ext_raw = ext_raw[ 0 : -1 ]

                        if pos_list and len( pos_list ) > 0:
                            for pos in pos_list:
                                tmp = re.sub( '\s+', ' ', pos.text.strip().replace( ',', ' ' ) ).split()
                                ext_pol.append( [ float(  tmp[ 0 ] ), float(  tmp[ 1 ] ) ] )
                                ext_raw += '%s %s,' % ( str( float(  tmp[ 1 ] ) ), str( float(  tmp[ 0 ] ) ) )
                            # remove last comma
                            ext_raw = ext_raw[ 0 : -1 ]
                        ext_raw += ')'

                        polygon_tmp.extend( ext_pol )
                        raw_tmp += ext_raw

                if interior:
                    ring = int_.findall( '{http://www.opengis.net/gml}LinearRing' )

                    if ring:
                        poslist = ring.find( '{http://www.opengis.net/gml}posList' )
                        # get all pos nodes, if existing
                        pos_list = ring.findall( './/{http://www.opengis.net/gml}pos' )

                        int_pol = []
                        int_raw = '('
                        if poslist:
                            tmp = re.sub( '\s+', ' ', poslist.text.strip().replace( ',', ' ' ) ).split()
                            # Gel all Latitudes
                            lat_list = tmp[ 0 : len( tmp ) : 2 ]
                            # Gel all Longitudes
                            lon_list = tmp[ 1 : len( tmp ) : 2 ]

                            for lat, lon in zip( lat_list, lon_list ):
                                int_pol.append( [ float( lat ), float( lon ) ] )
                                int_raw += '%s %s,' % ( str( float( lon ) ), str( float( lat ) ) )
                            # remove last comma
                            int_raw = int_raw[ 0 : -1 ]

                        if pos_list and len( pos_list ) > 0:
                            for pos in pos_list:
                                tmp = re.sub( '\s+', ' ', pos.text.strip().replace( ',', ' ' ) ).split()
                                int_pol.append( [ float(  tmp[ 0 ] ), float(  tmp[ 1 ] ) ] )
                                int_raw += '%s %s,' % ( str( float(  tmp[ 1 ] ) ), str( float(  tmp[ 0 ] ) ) )
                            # remove last comma
                            int_raw = int_raw[ 0 : -1 ]
                        int_raw += ')'

                        polygon_tmp.extend( int_pol )
                        raw_tmp += int_raw

                self.polygon_list.append( {
                    'type': 'POLYGON',
                    'mode': 'latlon',
                    'coordinates': polygon_tmp,
                    'raw': 'POLYGON(%s)' % raw_tmp,
                } )
            return True

        logger.debug( '  No "GML Polygon" detected! Nothing to decode.' )
        return False

    def __gml_boxes( self ):
        '''
        Example:
            <gml:Envelope>
                <gml:lowerCorner>42.943 -71.032</gml:lowerCorner>
                <gml:upperCorner>43.039 -69.856</gml:upperCorner>
            </gml:Envelope>

        The following are DEPRECATED:
            <gml:Envelope>
                <gml:coord>...</gml:coord>
                <gml:coord>...</gml:coord>
            </gml:Envelope>

            <gml:Envelope>
                <gml:pos>...</gml:pos>
                <gml:pos>...</gml:pos>
            </gml:Envelope>

            <gml:Envelope>
                <gml:coordinates>...</gml:coordinates>
            </gml:Envelope>
        '''

        envelope_list = self.where_node.findall( '{http://www.opengis.net/gml}Envelope' )
        if envelope_list:
            logger.debug( '  "GML Box" detected! Decoding...' )
            for envelope in envelope_list:
                lc = envelope.find( '{http://www.opengis.net/gml}lowerCorner' )
                lc_raw = re.sub( '\s+', ' ', lc.text.strip().replace( ',', ' ' ) )
                lc_tmp = lc_raw.split()

                uc = envelope.find( '{http://www.opengis.net/gml}upperCorner' )
                uc_raw = re.sub( '\s+', ' ', uc.text.strip().replace( ',', ' ' ) )
                uc_tmp = uc_raw.split()

                polygon_tmp = []
                polygon_tmp.append( [ float( uc_tmp[ 0 ] ), float( uc_tmp[ 1 ] ) ] )
                polygon_tmp.append( [ float( lc_tmp[ 0 ] ), float( uc_tmp[ 1 ] ) ] )
                polygon_tmp.append( [ float( lc_tmp[ 0 ] ), float( lc_tmp[ 1 ] ) ] )
                polygon_tmp.append( [ float( uc_tmp[ 0 ] ), float( lc_tmp[ 1 ] ) ] )
                polygon_tmp.append( [ float( uc_tmp[ 0 ] ), float( uc_tmp[ 1 ] ) ] )
                raw_tmp = ''
                raw_tmp += '%s %s,' % ( str( float( uc_tmp[ 1 ] ) ), str( float( uc_tmp[ 0 ] ) ) )
                raw_tmp += '%s %s,' % ( str( float( uc_tmp[ 1 ] ) ), str( float( lc_tmp[ 0 ] ) ) )
                raw_tmp += '%s %s,' % ( str( float( lc_tmp[ 1 ] ) ), str( float( lc_tmp[ 0 ] ) ) )
                raw_tmp += '%s %s,' % ( str( float( lc_tmp[ 1 ] ) ), str( float( uc_tmp[ 0 ] ) ) )
                raw_tmp += '%s %s'  % ( str( float( uc_tmp[ 1 ] ) ), str( float( uc_tmp[ 0 ] ) ) )
                self.polygon_list.append( {
                    'type': 'POLYGON',
                    'mode': 'latlon',
                    'coordinates': [ polygon_tmp ],
                    'raw': 'POLYGON((%s))' % raw_tmp,
                } )
            return True

        logger.debug( '  No "GML Box" detected! Nothing to decode.' )
        return False
